File: abstractions/dotnet/src/serialization/SerializationWriterFactoryRegistry.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;

namespace Microsoft.Kiota.Abstractions.Serialization
{
    /// <summary>
    ///  This factory holds a list of all the registered factories for the various types of nodes.
    /// </summary>
    public class SerializationWriterFactoryRegistry : ISerializationWriterFactory
    {
        /// <summary>
        /// The valid content type for the <see cref="SerializationWriterFactoryRegistry"/>
        /// </summary>
        public string ValidContentType
        {
            get
            {
                throw new InvalidOperationException("The registry supports multiple content types. Get the registered factory instead.");
            }
        }
        /// <summary>
        /// Default singleton instance of the registry to be used when registring new factories that should be available by default.
        /// </summary>
        public static readonly SerializationWriterFactoryRegistry DefaultInstance = new();
        /// <summary>
        /// List of factories that are registered by content type.
        /// </summary>
        public Dictionary<string, ISerializationWriterFactory> ContentTypeAssociatedFactories { get; set; } = new Dictionary<string, ISerializationWriterFactory>();
        /// <summary>
        /// Get the relevant <see cref="ISerializationWriter"/> instance for the given content type
        /// </summary>
        /// <param name="contentType">The content type in use</param>
        /// <returns>A <see cref="ISerializationWriter"/> instance to parse the content</returns>
        public ISerializationWriter GetSerializationWriter(string contentType)
        {
            if(string.IsNullOrEmpty(contentType))
                throw new ArgumentNullException(nameof(contentType));

            if(ContentTypeAssociatedFactories.ContainsKey(contentType))
                return ContentTypeAssociatedFactories[contentType].GetSerializationWriter(contentType);
            else
                throw new InvalidOperationException($"Content type {contentType} does not have a factory registered to be parsed");
        }

    }
}

File: abstractions/php/src/Serialization/SerializationWriterFactoryRegistry.php
<?php


namespace Microsoft\Kiota\Abstractions\Serialization;


use InvalidArgumentException;
use RuntimeException;
use UnexpectedValueException;

class SerializationWriterFactoryRegistry implements SerializationWriterFactory {

    /**
     * List of factories that are registered by content type
     * @var array<string, SerializationWriterFactory>
     */
    public array $contentTypeAssociatedFactories = [];

    /**
     * Default singleton instance of the registry to be used when registering new factories that should be available by default.
     * @var SerializationWriterFactoryRegistry|null
     */
    private static ?SerializationWriterFactoryRegistry $defaultInstance = null;

    private const VENDOR_SPECIFIC_CONTENT_TYPE_PATTERN = "/[^\/]+\\+/";

    /**
     * @param string $contentType
     * @return SerializationWriter
     * @throws UnexpectedValueException
     */
    public function getSerializationWriter(string $contentType): SerializationWriter {
        if (trim($contentType) === '') {
            throw new InvalidArgumentException('contentType cannot be empty');
        }

        $vendorSpecificContentType = explode(';', $contentType)[0];

        if (array_key_exists($vendorSpecificContentType, $this->contentTypeAssociatedFactories)) {
            return $this->contentTypeAssociatedFactories[$vendorSpecificContentType]->getSerializationWriter($vendorSpecificContentType);
        }

        $cleanedContentType = preg_replace(self::VENDOR_SPECIFIC_CONTENT_TYPE_PATTERN, '', $vendorSpecificContentType);

        if ($cleanedContentType === null) {
            throw new UnexpectedValueException("An error occurred processing the vendor specific content type.");
        }

        if (array_key_exists($cleanedContentType, $this->contentTypeAssociatedFactories)) {
            return $this->contentTypeAssociatedFactories[$cleanedContentType]->getSerializationWriter($cleanedContentType);
        }
        throw new UnexpectedValueException('Content type ' . $contentType . ' does not have a factory to be parsed');
    }

    public function getValidContentType(): string {
        throw new RuntimeException("The registry supports multiple content types. Get the registered factory instead.");
    }

    /**
     * Gets the singleton default instance of  @link SerializationWriterFactoryRegistry
     * @return SerializationWriterFactoryRegistry
     */
    public static function getDefaultInstance(): SerializationWriterFactoryRegistry {
        if (is_null(self::$defaultInstance)) {
            self::$defaultInstance = new self();
        }
        return self::$defaultInstance;
    }
}